### ISSUE TITLE
Issue with --prefer-lowest

### ISSUE THREAD

**Issue by Toflar** [composer/composer#6476]
My `composer.json`:

```json
{
    "name": "foobar/test",
    "type": "project",
    "require": {
        "friendsofsymfony/http-cache": "^2.0",
        "php-http/guzzle6-adapter": "^1.1"
    }
}
```

When I run this command:

```
composer update --prefer-lowest
```

I get the following output:

```
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for friendsofsymfony/http-cache ^2.0 -> satisfiable by friendsofsymfony/http-cache[2.0.0].
    - friendsofsymfony/http-cache 2.0.0 requires php-http/client-common ^1.1.0 -> satisfiable by php-http/client-common[1.3.0, 1.5.0, v1.1.0, v1.2.0, v1.2.1, v1.4.0, v1.4.1, v1.4.2].
    - Conclusion: don't install php-http/client-common v1.4.2
    - Conclusion: don't install php-http/client-common v1.4.1
    - Conclusion: don't install php-http/client-common v1.4.0
    - Conclusion: don't install php-http/client-common 1.3.0
    - Conclusion: don't install php-http/client-common v1.2.1
    - Conclusion: don't install php-http/client-common v1.2.0
    - Conclusion: don't install php-http/client-common v1.1.0
    - php-http/client-common 1.5.0 requires php-http/message ^1.2 -> satisfiable by php-http/message[1.5.0, v1.2.0, v1.3.0, v1.3.1, v1.4.0, v1.4.1].
    - php-http/message 1.5.0 requires php-http/message-factory ^1.0.2 -> satisfiable by php-http/message-factory[v1.0.2].
    - php-http/message v1.2.0 requires php-http/message-factory ^1.0.2 -> satisfiable by php-http/message-factory[v1.0.2].
    - php-http/message v1.3.0 requires php-http/message-factory ^1.0.2 -> satisfiable by php-http/message-factory[v1.0.2].
    - php-http/message v1.3.1 requires php-http/message-factory ^1.0.2 -> satisfiable by php-http/message-factory[v1.0.2].
    - php-http/message v1.4.0 requires php-http/message-factory ^1.0.2 -> satisfiable by php-http/message-factory[v1.0.2].
    - php-http/message v1.4.1 requires php-http/message-factory ^1.0.2 -> satisfiable by php-http/message-factory[v1.0.2].
    - Conclusion: don't install php-http/message-factory v1.0.2
```

And I expected this to happen:

I expected it to resolve to a valid set of dependencies :-) Because of `--prefer-lowest` I think it should take `php-http/message v1.2.0`, of which Composer tells me `php-http/message v1.2.0 requires php-http/message-factory ^1.0.2 -> satisfiable by php-http/message-factory[v1.0.2]`. However, the conclusion is `Conclusion: don't install php-http/message-factory v1.0.2` even though it is satisfiable?

When I run `composer update`, everything runs fine and all dependencies are resolved. Funnily enough, when I run `composer update` first, and the `vendor` dir exists (specifically speaking, the `installed.json`), all of a sudden `composer update --prefer-lowest` works too!

I could temporarily fix it by using this `composer.json`:

```json
{
    "name": "foobar/test",
    "type": "project",
    "require": {
        "friendsofsymfony/http-cache": "^2.0",
        "php-http/guzzle6-adapter": "^1.1"
    },
    "require-dev": {
        "php-http/message-factory": "^1.0.2"
    }
}
```

**Comment by ausi** [comment]
It looks like another temporarily fix is with a conflict:

```json
"conflict": {
    "php-http/message-factory": "< 1.0.2"
}
```

**Comment by alcohol** [comment]
Something seems off here, because your example composer.json does not contain any constraints, only specific hardcoded versions, `--prefer-lower` would have zero effect.

Please share actually reproducible scenarios.

**Comment by Toflar** [comment]
Sorry, fixed the example. It was actually correct the second time I mentioned it (with the temporary fix). It is easily reproducible. Just copy it and run with `--prefer-lowest`. It will fail.

**Comment by ausi** [comment]
I think `--prefer-lowest` prefers the lowest version of all dependencies, including dependencies of dependencies:

- `friendsofsymfony/http-cache 2.0.0` requires `php-http/client-common ^1.1.0`
- `php-http/client-common 1.1.0` requires `php-http/message ^1.2`
- `php-http/message v1.2.0` requires `php-http/message-factory ^1.0.2`

So even with fixed versions in the main *composer.json* file `--prefer-lowest` still has an effect and in this case causes the issue.

You can reproduce the issue with a single command, execute the following in an empty directory:

```sh
composer require --prefer-lowest friendsofsymfony/http-cache:2.0.0 php-http/guzzle6-adapter:1.1.0
```

**Comment by dereuromark** [comment]
Sounds similar to https://github.com/composer/composer/issues/7849 somewhat

**Comment by naderman** [comment]
Should be fixed with https://github.com/composer/composer/pull/7946 - please verify if you can.

**Comment by ausi** [comment]
With the original set of packages the issue was not reproducible on my system anymore, but with `sebastian/diff:^1.4` and `phpunit/phpunit:^5.7.26` it still is.

For the following command:

```sh
composer require --prefer-lowest sebastian/diff:^1.4 phpunit/phpunit:^5.7.26
```

- Composer 1.7.2 shows: `Your requirements could not be resolved to an installable set of packages.`
- Composer 1.8.3 shows: `Your requirements could not be resolved to an installable set of packages.`
- Composer from #7946 (b09a4de08b15efa85b496471c074ccac72d65ddb) shows: `Package operations: 23 installs, 0 updates, 0 removals`

So it looks like the fix from #7946 works for this kind of issue 🎉. Thank you @naderman!

**Comment by naderman** [comment]
You can use a current composer snapshot to test this `composer self-update --snapshot`.

I'll close this because I think it was fixed. Please comment/reopen if this is still a problem.
